include/systems/eigen_system.h

```diff
@@ -175,56 +165,261 @@ class EigenSystem : public System
+    {
+      libmesh_assert(precond_matrix);
+      libmesh_assert(!_use_shell_matrices);
+      libmesh_assert_equal_to(&get_matrix(precond_matrix_name()), precond_matrix);
+      return *precond_matrix;
+    }
+  /**
+   * \returns A reference to the preconditioning matrix.
+   *
+   * This matrix should only be available (and therefore this should only be called)
+   * if !_use_shell_matrices.
+   */
+  inline SparseMatrix<Number> & get_precond_matrix()
+    {
+      libmesh_assert(precond_matrix);
+      libmesh_assert(!_use_shell_matrices);
+      libmesh_assert_equal_to(&get_matrix(precond_matrix_name()), precond_matrix);
+      return *precond_matrix;
+    }
+
+  /**
+   * \returns A const reference to the system shell matrix used for standard eigenvalue problems.
+   *
+   * This matrix should only be available (and therefore this should only be called)
+   * if _use_shell_matrices.
+   */
+  inline const ShellMatrix<Number> & get_shell_matrix_A() const
+    {
+      libmesh_assert(_use_shell_matrices);
+      libmesh_assert(shell_matrix_A);
+      return *shell_matrix_A;
+    }
+  /**
+   * \returns A reference to the system shell matrix used for standard eigenvalue problems.
+   *
+   * This matrix should only be available (and therefore this should only be called)
+   * if _use_shell_matrices.
+   */
+  inline ShellMatrix<Number> & get_shell_matrix_A()
+    {
+      libmesh_assert(_use_shell_matrices);
+      libmesh_assert(shell_matrix_A);
+      return *shell_matrix_A;
+    }
+
+  /**
+   * \returns A const reference to the system shell matrix used for generalized eigenvalue problems.
+   *
+   * This matrix should only be available (and therefore this should only be called)
+   * if _use_shell_matrices and generalized().
+   */
+  inline const ShellMatrix<Number> & get_shell_matrix_B() const
+    {
+      libmesh_assert(_use_shell_matrices);
+      libmesh_assert(generalized());
+      libmesh_assert(shell_matrix_B);
+      return *shell_matrix_B;
+    }
+  /**
+   * \returns A reference to the system shell matrix used for generalized eigenvalue problems.
+   *
+   * This matrix should only be available (and therefore this should only be called)
+   * if _use_shell_matrices and generalized().
+   */
+  inline ShellMatrix<Number> & get_shell_matrix_B()
+    {
+      libmesh_assert(_use_shell_matrices);
+      libmesh_assert(generalized());
+      libmesh_assert(shell_matrix_B);
+      return *shell_matrix_B;
+    }
+
+  /**
+   * \returns A const reference to the system shell matrix used for preconditioning.
+   */
+  inline const ShellMatrix<Number> & get_shell_precond_matrix() const
+    {
+      libmesh_assert(_use_shell_matrices);
+      libmesh_assert(_use_shell_precond_matrix);
+      libmesh_assert(shell_precond_matrix);
+      return *shell_precond_matrix;
+    }
+  /**
+   * \returns A reference to the system shell matrix used for preconditioning.
+   */
+  inline ShellMatrix<Number> & get_shell_precond_matrix()
```

Comment:
I guess somewhere we have a function:

`bool has_shell_precond_matrix?`

Reply:
`EigenSystem::use_shell_precond_matrix()` and `EigenSystem::use_shell_precond_matrix(bool)`

Reply:
Oh, I was wondering how to have a safe way to check if we have a shell matrix. Right now, I directly access the pointer

Reply:
I have a PR ready in MOOSE that will do safe checking and use the getter methods instead of directly accessing the public pointers (which will be deprecated in the future).